### ISSUE TITLE
[🐛 Bug]: Custom capability is not working properly with IE in selenium 4

### ISSUE THREAD

**Issue by mpstellar** [SeleniumHQ/selenium#11625]
### What happened?

I have a hub with two kind of different Grid nodes. I am trying to set a custom-cap in IE (Edge with Internet Explorer Mode), however, when I set (“custom-cap”, false), the tests are executed on any node indistinctly.
I have trying to set different kind of prefix as well, but nothing works: (“c:custom-cap”, false), (“ie:custom-cap”, false), (“se:custom-cap”, false).

### How can we reproduce the issue?

```shell
The custom capability is applied on the http://hubipadress:4444/status page:
{
  "value": {
    "ready": true,
    "message": "Selenium Grid ready.",
    "nodes": [
      {
        "id": "8c748cc5-a767-4392-b0d7-4abffc16aec1",
        "uri": "http:\u002f\u002f.....:5555",
        "maxSessions": 2,
        "osInfo": {
          "arch": "amd64",
          "name": "Windows Server 2016",
          "version": "10.0"
        },
        "heartbeatPeriod": 60000,
        "availability": "UP",
        "version": "4.7.2 (revision 4d4020c3b7)",
        "slots": [
          {
            "id": {
              "hostId": "8c748cc5-a767-4392-b0d7-4abffc16aec1",
              "id": "37ef5819-e892-471e-9597-bb5020b5cbab"
            },
            "lastStarted": "1970-01-01T00:00:00Z",
            "session": null,
            "stereotype": {
              "browserName": "chrome",
              "custom-cap": false,
              "platformName": "Windows 10",
              "se:webDriverExecutable": "D:\\Selenium\\Drivers\\chromedriver.exe"
            }
          },
          {
            "id": {
              "hostId": "8c748cc5-a767-4392-b0d7-4abffc16aec1",
              "id": "e549d761-5f80-4acb-8abb-dc6a11452f8b"
            },
            "lastStarted": "1970-01-01T00:00:00Z",
            "session": null,
            "stereotype": {
              "browserName": "firefox",
              "custom-cap": false,
              "platformName": "Windows 10",
              "se:webDriverExecutable": "D:\\Selenium\\Drivers\\geckodriver.exe"
            }
          },
          {
            "id": {
              "hostId": "8c748cc5-a767-4392-b0d7-4abffc16aec1",
              "id": "f0351165-16bf-4ca1-8a01-3e78b455e82d"
            },
            "lastStarted": "1970-01-01T00:00:00Z",
            "session": null,
            "stereotype": {
              "browserName": "internet explorer",
              "custom-cap": false,
              "platformName": "Windows 10",
              "se:ieOptions": {
                "ie.edgechromium": "true",
                "ie.edgepath": "C:\u002fProgram Files (x86)\u002fMicrosoft\u002fEdge\u002fApplication\u002fmsedge.exe"
              },
              "se:webDriverExecutable": "D:\\Selenium\\Drivers\\IEDriverServer.exe"
            }
          },
          {
            "id": {
              "hostId": "8c748cc5-a767-4392-b0d7-4abffc16aec1",
              "id": "285ba454-0da4-4192-a5c4-1e99f3049bcb"
            },
            "lastStarted": "1970-01-01T00:00:00Z",
            "session": null,
            "stereotype": {
              "browserName": "MicrosoftEdge",
              "custom-cap": false,
              "platformName": "Windows 10",
              "se:webDriverExecutable": "D:\\Selenium\\Drivers\\msedgedriver.exe"
            }
          },
          {
            "id": {
              "hostId": "8c748cc5-a767-4392-b0d7-4abffc16aec1",
              "id": "9c79b70c-f375-4862-b06e-8c00de974b7b"
            },
            "lastStarted": "1970-01-01T00:00:00Z",
            "session": null,
            "stereotype": {
              "browserName": "firefox",
              "custom-cap": false,
              "platformName": "Windows 10",
              "se:webDriverExecutable": "D:\\Selenium\\Drivers\\geckodriver.exe"
            }
          },
          {
            "id": {
              "hostId": "8c748cc5-a767-4392-b0d7-4abffc16aec1",
              "id": "bf6c3818-6042-482a-9d5b-0ecf34c29e1b"
            },
            "lastStarted": "1970-01-01T00:00:00Z",
            "session": null,
            "stereotype": {
              "browserName": "chrome",
              "custom-cap": false,
              "platformName": "Windows 10",
              "se:webDriverExecutable": "D:\\Selenium\\Drivers\\chromedriver.exe"
            }
          },
          {
            "id": {
              "hostId": "8c748cc5-a767-4392-b0d7-4abffc16aec1",
              "id": "8a3fba1d-0747-45fb-8ad3-32c458ed35a1"
            },
            "lastStarted": "1970-01-01T00:00:00Z",
            "session": null,
            "stereotype": {
              "browserName": "MicrosoftEdge",
              "custom-cap": false,
              "platformName": "Windows 10",
              "se:webDriverExecutable": "D:\\Selenium\\Drivers\\msedgedriver.exe"
            }
          }
        ]
      },
      {
        "id": "5ce2a7c7-1a56-44d9-8391-e9ba9a2ff3fe",
        "uri": "http:\u002f\u002f10.151.18.98:5555",
        "maxSessions": 2,
        "osInfo": {
          "arch": "amd64",
          "name": "Linux",
          "version": "3.10.0-1160.71.1.el7.x86_64"
        },
        "heartbeatPeriod": 60000,
        "availability": "UP",
        "version": "4.7.2 (revision 4d4020c3b7)",
        "slots": [
          {
            "id": {
              "hostId": "5ce2a7c7-1a56-44d9-8391-e9ba9a2ff3fe",
              "id": "16846465-ec7e-4a66-96d2-7946e4f194cf"
            },
            "lastStarted": "1970-01-01T00:00:00Z",
            "session": null,
            "stereotype": {
              "browserName": "firefox",
              "custom-cap": false,
              "platformName": "LINUX",
              "se:webDriverExecutable": "\u002fapp\u002fselenium-connector\u002fgeckoDriver"
            }
          },
          {
            "id": {
              "hostId": "5ce2a7c7-1a56-44d9-8391-e9ba9a2ff3fe",
              "id": "27100db9-b8a0-4b26-9f85-7922835c8c56"
            },
            "lastStarted": "1970-01-01T00:00:00Z",
            "session": null,
            "stereotype": {
              "browserName": "firefox",
              "custom-cap": false,
              "platformName": "LINUX",
              "se:webDriverExecutable": "\u002fapp\u002fselenium-connector\u002fgeckoDriver"
            }
          },
          {
            "id": {
              "hostId": "5ce2a7c7-1a56-44d9-8391-e9ba9a2ff3fe",
              "id": "06e3caea-c951-4155-8dde-6f3e379abf67"
            },
            "lastStarted": "2023-02-07T08:51:00.449534Z",
            "session": null,
            "stereotype": {
              "browserName": "chrome",
              "custom-cap": false,
              "platformName": "LINUX",
              "se:webDriverExecutable": "\u002fapp\u002fselenium-connector\u002fchromeDriver"
            }
          },
          {
            "id": {
              "hostId": "5ce2a7c7-1a56-44d9-8391-e9ba9a2ff3fe",
              "id": "0626fe53-7d75-4631-86f5-1ddbdcebcc1b"
            },
            "lastStarted": "2023-02-07T08:53:13.896321Z",
            "session": null,
            "stereotype": {
              "browserName": "chrome",
              "custom-cap": false,
              "platformName": "LINUX",
              "se:webDriverExecutable": "\u002fapp\u002fselenium-connector\u002fchromeDriver"
            }
          }
        ]
      },
      {
        "id": "cf0e8b1b-7b0c-4b51-a10d-c45ed9af2cbd",
        "uri": "http:\u002f\u002f.....:5555",
        "maxSessions": 1,
        "osInfo": {
          "arch": "amd64",
          "name": "Windows Server 2016",
          "version": "10.0"
        },
        "heartbeatPeriod": 60000,
        "availability": "UP",
        "version": "4.7.2 (revision 4d4020c3b7)",
        "slots": [
          {
            "id": {
              "hostId": "cf0e8b1b-7b0c-4b51-a10d-c45ed9af2cbd",
              "id": "0ebb53a1-01b9-4486-8892-2b76903a7a3f"
            },
            "lastStarted": "1970-01-01T00:00:00Z",
            "session": null,
            "stereotype": {
              "browserName": "internet explorer",
              "custom-cap": true,
              "platformName": "Windows 10",
              "se:ieOptions": {
                "ie.edgechromium": "true",
                "ie.edgepath": "C:\u002fProgram Files (x86)\u002fMicrosoft\u002fEdge\u002fApplication\u002fmsedge.exe"
              },
              "se:webDriverExecutable": "D:\\Selenium\\Drivers\\IEDriverServer.exe"
            }
          }
        ]
      }
    ]
  }
}

I am just interested in IE browser. Bellow you can find part of th code which it´s running:
public InternetExplorerCaixa(boolean cap){
	options = new InternetExplorerOptions();
	options.attachToEdgeChrome();
	options.withEdgeExecutablePath("C:\\Program Files (x86)\\Microsoft\\Edge\\Application\\msedge.exe");
	if (cap) {
		options.setCapability(custom-cap, true);
	} else {
		options.setCapability(custom-cap, false);
	}
}
```

### Relevant log output

```shell
The expected behavior is that tests without custom-cap run just on the node where custom-cap is false, unfortunately they run on all of them.
```

### Operating System

Windows 10

### Selenium version

Selenium Server 4.7.2

### What are the browser(s) and version(s) where you see this issue?

Internet Explore 11

### What are the browser driver(s) and version(s) where you see this issue?

IEDriverServer.exe 4.7.0.0 (32-bit)

### Are you using Selenium Grid?

4.7.2

**Comment by diemol** [comment]
`"custom-cap": false,` in the stereotype needs to be prefixed, but please avoid `se:`.

It could be `"prefix:custom-cap": false,`.

**Comment by mpstellar** [comment]
I was trying with "c:cutom-cap": false and "ie:custom-cap": fase, but nothing works

**Comment by diemol** [comment]
How does the Grid look like when you set the custom cap with a prefix?

**Comment by mpstellar** [comment]
The Grid recognizes the custom-cap. Log output:
13:21:31  org.openqa.selenium.SessionNotCreatedException: Could not start a new session. Response code 500. Message: Could not start a new session. Could not start a new session. Error while creating session with the driver service. Stopping driver service: Could not start a new session. Response code 500. Message: No matching capability sets found.
13:21:31  Unable to match capability set 0: ie:custom-cap is an unknown extension capability for IE 
*****************************************************************************************************************
12:33:17  org.openqa.selenium.SessionNotCreatedException: Could not start a new session. Response code 500. Message: Could not start a new session. Could not start a new session. Error while creating session with the driver service. Stopping driver service: Could not start a new session. Response code 500. Message: No matching capability sets found.
12:33:17  Unable to match capability set 0: c:custom-cap is an unknown extension capability for IE

**Comment by diemol** [comment]
What I meant is how `http://hubipadress:4444/status` looks when you set the custom cap in the stereotype.

**Comment by mpstellar** [comment]
``` 
{
            "id": {
              "hostId": "8c748cc5-a767-4392-b0d7-4abffc16aec1",
              "id": "f0351165-16bf-4ca1-8a01-3e78b455e82d"
            },
            "lastStarted": "1970-01-01T00:00:00Z",
            "session": null,
            "stereotype": {
              "browserName": "internet explorer",
              "c:custom-cap": false,
              "platformName": "Windows 10",
              "se:ieOptions": {
                "ie.edgechromium": "true",
                "ie.edgepath": "C:\u002fProgram Files (x86)\u002fMicrosoft\u002fEdge\u002fApplication\u002fmsedge.exe"
              },
              "se:webDriverExecutable": "D:\\Selenium\\Drivers\\IEDriverServer.exe"
            }
          }, {
            "id": {
              "hostId": "8c748cc5-a767-4392-b0d7-4abffc16aec1",
              "id": "f0351165-16bf-4ca1-8a01-3e78b455e82d"
            },
            "lastStarted": "1970-01-01T00:00:00Z",
            "session": null,
            "stereotype": {
              "browserName": "internet explorer",
              "c:custom-cap": false,
              "platformName": "Windows 10",
              "se:ieOptions": {
                "ie.edgechromium": "true",
                "ie.edgepath": "C:\u002fProgram Files (x86)\u002fMicrosoft\u002fEdge\u002fApplication\u002fmsedge.exe"
              },
              "se:webDriverExecutable": "D:\\Selenium\\Drivers\\IEDriverServer.exe"
            }
          },
```

**Comment by diemol** [comment]
Does it work when you do not use the custom caps in the stereotype?

**Comment by mpstellar** [comment]
Yes, it does. Tests run properly in IE without any custom cap in the stereotype

**Comment by diemol** [comment]
OK, then please provide the steps you follow to start the Grid, with commands please.

And then the code you are using on the client side to create the session. Ideally we can reopen with that information.

**Comment by mpstellar** [comment]
I am working on a hub-node grid configuration similar to Selenium 3:

**Starting the HUB (LINUX)**

nohup -jar path/selenium-server-4.7.2.jar hub --config path/hubConfig.toml > path/nohup_selenium.out 2>&1 &

hubConfig.toml
```toml
title = "Configurar Selenium 4 en modo HUB"

[server]
        registration-secret = "clave"
        host = "localhost"
        port = 4444
        allow-cors = true
        bind-host = false

[sessionqueue]
        host = "http://localhost:1234"
        hostname = "localhost"
        port = 1234
        #18h
        session-request-timeout = 64800
        session-retry-interval = 5

[distributor]
        host = "http://localhost:5553"
        hostname = "localhost"
        port = 5553
        healthcheck-interval = 120
        reject-unsupported-caps = true


[events]
        implementation = "org.openqa.selenium.events.zeromq.ZeroMqEventBus"
        #publish = "tcp://*:4442"
        #subscribe = "tcp://*:4443"
        publish = "tcp://*:5558"
        subscribe = "tcp://*:5559"
        bind = true


[network]
        relax-checks = false
```		

 **Starting NODE 1(WINDOWS)**

java -jar path/selenium-server-4.7.2.jar node --config path/node1.toml  1>path/selenium-connector-node1.out 2>&1


node1.toml
```toml
title = "Configurar Selenium 4 en modo Node"

[server]
    registration-secret = "clave"
    host = "node1IPaddress"
    port = 5555
    allow-cors = true
    bind-host = true


[node]
    detect-drivers = false
    drain-after-session-count = 0

    [[node.driver-configuration]]
        display-name = "Internet Explorer"
        webdriver-executable = "path\\IEDriverServer.exe"
        max-sessions = 1
        stereotype = "{\"browserName\": \"internet explorer\", \"c:custom-cap\":  false, \"platformName\": \"WIN10\", \"se:ieOptions\": {\"ie.edgechromium\":\"true\", \"ie.edgepath\": \"C:/Program Files (x86)/Microsoft/Edge/Application/msedge.exe\"}}"

    [[node.driver-configuration]]
        display-name = "Firefox"
        webdriver-executable = "path\\geckodriver.exe"
        max-sessions = 5
        stereotype = "{\"browserName\": \"firefox\", \"c:custom-cap\":  false, \"platformName\": \"WIN10\"}"

    [[node.driver-configuration]]
        display-name = "Chrome"
        webdriver-executable = "path\\chromedriver.exe"
        max-sessions = 5
        stereotype = "{\"browserName\": \"chrome\", \"c:custom-cap\":  false, \"platformName\": \"WIN10\"}"

    [[node.driver-configuration]]
        display-name = "Edge"
        webdriver-executable = "path\\msedgedriver.exe"
        max-sessions = 5
        stereotype = "{\"browserName\": \"MicrosoftEdge\", \"c:custom-cap\":  false, \"platformName\": \"WIN10\"}"


    implementation = "org.openqa.selenium.grid.node.local.LocalNodeFactory"
    hub = "http://hubIPaddress:4444"
    grid-url = "http://hubIPaddress:4444"
    enable-cdp = true
    heartbeat-period = 60

    register-cycle = 10
    register-period = 120
    session-timeout = 300

[events]
    bind = false
    publish = "tcp://hubIPaddress:5558"
    subscribe = "tcp://hubIPaddress:5559"
``` 

**Starting NODE 2 (WINDOWS)**

java -jar path/selenium-server-4.7.2.jar node --config path/node2.toml  1>path/selenium-connector-node2.out 2>&1  

```toml
node2.toml
title = "Configurar Selenium 4 en modo Node"

[server]
    registration-secret = "clave"
    host = "node2IPaddress"
    port = 5555
    allow-cors = true
    bind-host = true
    max-threads = 20


[node]
    detect-drivers = false
    drain-after-session-count = 0

    [[node.driver-configuration]]
        display-name = "Internet Explorer"
        webdriver-executable = "path\\IEDriverServer.exe"
        max-sessions = 1
        stereotype = "{\"browserName\": \"internet explorer\", \"c:custom-cap\":  true, \"platformName\": \"WIN10\", \"se:ieOptions\": {\"ie.edgechromium\":\"true\", \"ie.edgepath\": \"C:/Program Files (x86)/Microsoft/Edge/Application/msedge.exe\"}}"

    implementation = "org.openqa.selenium.grid.node.local.LocalNodeFactory"
    hub = "http://hubIPaddress:4444"
    grid-url = "http://hubIPaddress:4444"
    enable-cdp = true
    heartbeat-period = 60

    register-cycle = 10
    register-period = 120
    session-timeout = 300

[events]
    bind = false
    publish = "tcp://hubIPaddress:5558"
    subscribe = "tcp://hubIPaddress:5559"
```

**Sample code to match Node 1:**
```java
options = new InternetExplorerOptions();
options.setCapability("ie.edgechromium", true);
options.withEdgeExecutablePath("C:\\Program Files (x86)\\Microsoft\\Edge\\Application\\msedge.exe");
options.setCapability(custom-cap, false);
WebDriver driver = new RemoteWebDriver(new URL("http://hubIPaddress:4444"), options);
driver.get("aplicationToTestUrl");
	"some steps"
driver.quit();
```

**Sample code to match Node 2:**
```java
options = new InternetExplorerOptions();
options.setCapability("ie.edgechromium", true);
options.withEdgeExecutablePath("C:\\Program Files (x86)\\Microsoft\\Edge\\Application\\msedge.exe");
options.setCapability(custom-cap, true);
WebDriver driver = new RemoteWebDriver(new URL("http://hubIPaddress:4444"), options);
driver.get("aplicationToTestUrl");
	"some steps"
driver.quit();
```

**Comment by diemol** [comment]
Reopening for analysis.

**Comment by titusfortner** [comment]
If Internet Explorer is set to redirect to Edge on the machine, you shouldn't need to specify `"ie.edgechromium":"true", "ie.edgepath": "C:/Program Files (x86)/Microsoft/Edge/Application/msedge.exe` with the latest version of IE Driver. What if you just remove those?

**Comment by mpstellar** [comment]
I'll try it and let you know what's going on, thank you for your feedback

**Comment by mpstellar** [comment]
I have checked that Internet Explorer is not set to redirect to Edge on the machines

**Comment by titusfortner** [comment]
It should still automatically locate Edge, you would just need to specify the `"ie.edgechromium":"true"` if it might otherwise grab IE instead of Edge

**Comment by MajinKLS** [comment]
Hi. My coworker is taking sick leave, so I’ll assume this subject.
Basically, you are trying to validate if the redirection with custom capabilities on Edge compatibility mode is affected by the explicitly specification of Edge's path, right?

I'll check this in the following days.

Best regards.

**Comment by titusfortner** [comment]
I don't know how custom capabilities work with grid, I was trying to see if we could remove extra variables in case it would be easier to focus on toggling from the custom capability. I didn't mean to derail the conversation.

**Comment by diemol** [comment]
The issue was that the custom capability `c:custom-cap` was being rejected by the IEDriverServer. Ideally Grid should not forward those to the drivers because it can bring unexpected behaviors. For now, the Grid will remove custom caps when the browser is `internet explorer`, and based on feedback we might generalize the fix. 

This will be available in 4.9.0, which should be out next week.

**Comment by diemol** [comment]
Also, a few things, the Grid configuration had elements that were not necessary or were misleading. I simplified the configuration for this issue and it can be seen here https://github.com/diemol/selenium-issues/tree/master/selenium/issue-11625

You can also see the code to create the session https://github.com/diemol/selenium-issues/blob/master/selenium/issue-11625/src/main/java/org/example/Main.java#L16, and you do not need to set;

```
options.setCapability("ie.edgechromium", true);
options.withEdgeExecutablePath("C:\\Program Files (x86)\\Microsoft\\Edge\\Application\\msedge.exe");
```

in the code neither in the stereotype because those are the default values.

**Comment by MajinKLS** [comment]
Hi, sorry for the delay as I'm working on different blocking issues and I started this one from the begining.
As you said it is because IEDriverServer is rejecting custom capabilities. I found that last friday and I wanted to do more test, but at now our issue seems to be solved by using the test prefix, like "test:app", that I found in this other issue https://github.com/SeleniumHQ/selenium/issues/7479

So, on future version 4.9.0 custom caps wil not be avaliable with IE and test prefix will stop working?

Also I'll check your links.
